### PR TITLE
Reconnect Watch For ConfigMaps And Secrets

### LINKED ISSUE #557: Reload mode - event not working for k8 config - 1.1.2.RELEASE
I'm able to fetch the K8 config maps on startup but on subsequent update to the config map, its not resulting in consuming the event and updating the values in my spring boot application.

Here are the versions:
spring boot -     2.2.7.RELEASE    
spring-cloud-gcp - 1.2.2.RELEASE
spring-cloud - Hoxton.SR4

Its working  fine in 
spring boot -     2.1.11.RELEASE    
spring-cloud-gcp - 1.1.3.RELEASE
spring-cloud - Greenwich.SR4
Able to fetch the config map during startup as well as consuming the event when there is a change in the config map

### LINKED ISSUE #831: Config watcher not sending HTTP request after 404
I'm doing a poc of the spring config watcher. We set up the watcher with the yaml provided [here](https://docs.spring.io/spring-cloud-kubernetes/docs/current/reference/html/index.html#deployment-yaml) and are using an nginx pod as the test service that the watcher should ping to refresh. 

The first time the config map gets updated, the watcher logs show it sending the post request to the actuator endpoint and getting a 404 back. That's expected, but subsequent requests appear to not get sent at all. If I restart the config watcher, then it does send the request again, but if I remove the target app's pod and create a new one, it still doesn't make the post request.

It seems like it might be blacklisting that URL, which would make sense so the config watcher isn't constantly sending requests to an endpoint that doesn't exist, but I didn't see anything that seemed like it would be doing that. So I was wondering what the cause of this behavior is and if there's a way to trigger the watcher to send posts to the endpoint again without restarting it.

### LINKED ISSUE #837: Config watcher didn't response to configmap change after 45 minutes
**Describe the bug**
I'm using config watcher to enable auto refresh for my  spring cloud microservices deployed in aws EKS.
Config watcher is also deployed in EKS. 
It works fine when config watcher was just deployed, however, after 40 minutes, when I modified the configmap again to test it again, config watcher didn't response, no logs at all.

Config watcher version:  2.1.0-SNAPSHOT
Spring boot version: 2.4.4
JDK: 1.8.0_282

**Sample**
project link:  https://github.com/neo-feng/configdemo 
you may find all information from link above.

pom dependencies:
```
	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-actuator</artifactId>
		</dependency>
		<dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-starter-kubernetes-fabric8-config</artifactId>
            <version>${spring.cloud.config.version}</version>
        </dependency>
	</dependencies>
```

bootstrap.yml
```
spring:
  application:
    name: demo-service
  cloud:
    kubernetes:
      config:
        #enabled: true
        sources:
          - namespace: default
            name: ${spring.application.name}

server:
  port: 8080
```
reproduce steps:
- deploy the application and test it successfully according to readme.md in the link above
- after 40 minutes, test it again -- change the configmap and watch logs
- you'll see config watcher has no response, no logs at all

It seems config watcher was disconnected with kubernetes after 40 minutes, while it still works well after 20 minutes, has no idea how to proceed, any suggestion will be appreciated.

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Fixes #557
Fixes #837
Fixes #831

### CODE REVIEW

**Comment by ryanjbaxter** [comment]
@arobie1992 @neo-feng

There is a version of the config watcher image here in this public repo 
public.ecr.aws/c3y1x2r0/springcloud/spring-cloud-kubernetes-configuration-watcher:2.0.4-SNAPSHOT

Could you try it out and let me know if it fixes your issues?

File: spring-cloud-kubernetes-fabric8-config/src/main/java/org/springframework/cloud/kubernetes/fabric8/config/reload/EventBasedConfigMapChangeDetector.java [hunk 1]
```diff
@@ -84,7 +87,13 @@ public void eventReceived(Action action, ConfigMap configMap) {
...
 
 					@Override
-					public void onClose(KubernetesClientException e) {
+					public void onClose(KubernetesClientException exception) {
+						log.warn("ConfigMaps watch closed", exception);
+						Optional.ofNullable(exception).map(e -> {
+							log.debug("Exception received during watch", e);
+							return exception;
+						}).map(KubernetesClientException::getStatus).map(Status::getCode)
+								.filter(c -> c.equals(HttpURLConnection.HTTP_GONE)).ifPresent(c -> watch());
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-config/src/main/java/org/springframework/cloud/kubernetes/fabric8/config/reload/EventBasedConfigMapChangeDetector.java:96]: simple, but yet very nice. I was not aware that they respond with `HTTP_GONE`

**Comment by ryanjbaxter** [comment]
Forgot to add the SO post that describes this scenario https://stackoverflow.com/a/61437982

**Comment by neo-feng** [comment]
> @arobie1992 @neo-feng
> 
> There is a version of the config watcher image here in this public repo
> public.ecr.aws/c3y1x2r0/springcloud/spring-cloud-kubernetes-configuration-watcher:2.0.4-SNAPSHOT
> 
> Could you try it out and let me know if it fixes your issues?

I works. I tried twice, each waiting for 1 hour.

**Comment by arobie1992** [comment]
@ryanjbaxter Everything seems to be working for us as well. Thanks!

**Pull request merged**